Validate phone number before submit

arathjz opened the pull request:
**What:**
Validate phone number before submit

**Why:**
Relates to: https://app.asana.com/0/1159164196409156/1199565893187287/f

**How:**
- Add phone number validation to only enter numbers while the user is typing
- Add phone number length validation when the user submits the form

#### Media
![image](https://user-images.githubusercontent.com/20747535/114278665-7e9aa280-99f6-11eb-820a-2ee9cac60375.png)

orlando commented:
@arathjz there are a few issues with keyboard navigation on this specific step of the form. Do you think you can take a look on this same PR? It's ok also to work on this on another PR

orlando reviewed (approved):
LGTM

duranmla commented on packages/union-component/src/components/DonationPaymentForm.tsx:
```diff
@@ -44,13 +47,31 @@ const DonationPaymentForm: React.FC<Props> = ({
   const errorMessage: string | undefined = errors?.join(' ');
 
+  const isPhoneNumberValid = (phoneNumber: string) => {
+    const formattedNumber = phoneNumber.split(' ').join('');
```
Why we do this split/join, are we trying to do like `phoneNumber.replaceAll(' ', '')` kinda thing?

arathjz replied:
No specific reason, it's just another way to achieve the same result. I'd like more `replaceAll`

duranmla commented on packages/union-component/src/components/DonationPaymentForm.tsx:
```diff
@@ -155,6 +176,11 @@ const DonationPaymentForm: React.FC<Props> = ({
             {errorMessage || DEFAULT_ERROR}
           </DonationWizard.ErrorText>
         ) : null}
+        {phoneNumberError ? (
```
I would have expected to have the phone number error below the phone number input.
_Why are we adding the error here?_

Besides, I see some benefits into have state management within the phone input component. Because then this form component will only "known that the phone input will return a valid value"

duranmla commented on packages/union-component/src/components/DonationPhoneInput.tsx:
```diff
@@ -21,7 +23,7 @@ const DonationPhoneInput: React.FC<InputHTMLAttributes<HTMLInputElement>> = ({
 
   const handleOnChange = (phone: string) => {
-    setValue(phone);
+    if (NUMBER_REGEX.test(phone)) setValue(phone);
```
I think you can rather use `pattern` from `input`. https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern

arathjz replied:
@duranmla  The `react-phone-input-2` API doesn't allow to pass a `pattern` attribute

arathjz replied:
I found out there's no need to add this, the input already takes care of preventing the user typing letters or odd characters

duranmla reviewed (changes requested):
I would rather clarify couple doubts on the comments I made before merging.

orlando commented:
btw now we are checking this, the library we are using supports getting the phone number E.164 format. We are implementing this validation in the backend.

The other thing we need to change is removing the country code from the actual input, like this gif

http://g.recordit.co/NVfQRWktci.gif

arathjz commented:
@orlando I couldn't find a way to validate the E.164 format with the library, however, I added a validation before submiting to test we are using this format

```
const isPhoneNumberValid = (phoneNumber: string) => {
    const formattedNumber = phoneNumber.replaceAll(' ', '');

    return (
      E_164_PHONE_FORMAT.test(formattedNumber) &&
      formattedNumber.length >= PHONE_NUMBER_MIN_LENGTH
    );
  };
```

arathjz commented:
@orlando I just added a new commit with the E.164 validation. I wasn't able to hide the country code, where did you take the gif from? We may add a new ticket for this, I looked up in the `react-phone-input-2` docs and couldn't find a way to hide it

@duranmla I addressed your comments about the error we display to the user, I moved it below the phone input

orlando commented:
@arathjz The gif is from charity water, I think we need to find a way to hide the country code. If it's not possible with React or this library, we may try this one https://www.npmjs.com/package/react-phone-number-input. It gives you the phone number in in E.164 format
